### ISSUE TITLE
Move -O3 flag to FFLAGS

### ISSUE THREAD

**Issue by mcgibbon** [CliMT/climt#50]
Currently it is difficult to manually set, for example `-O0` for debugging. If it is included in FFLAGS in setup.py, then when setup.py runs it will be added to lines that already include `-O3`. These instances of `-O3` should be removed from those files and put in FFLAGS in setup.py instead.

**Comment by JoyMonteiro** [comment]
FFLAGS is a global flag, and we don't want it to be communicated through downstream to every compile command. The reason `-O3` is not global is because some files in RRTMG are compiled without this flag, and adding it increases compile time by quite a bit. 

I'm open to other options, like creating a `CLIMT_OPTIMIZE_FLAG=-O3` and using it in downstream Makefiles.

**Comment by mcgibbon** [comment]
CLIMT_OPTIMIZE_FLAG would be a good option to have.

**Comment by mcgibbon** [comment]
Does `-O3` affect RRTMG run times? I would imagine that's the module you most want to optimize.

**Comment by JoyMonteiro** [comment]
I'm sure it does, and the original RRTMG Makefile keeps this option on. However, there are huge data files containing the k-tables, which are compiled without this flag.

**Comment by mcgibbon** [comment]
I would imagine that if we want fast compilation, we should just use a lower `O` flag, and if we want fast code, we should use `-O3` even for RRTM. I don't see why RRTM should have a different optimization flag from the rest of the code.

**Comment by mcgibbon** [comment]
Like, think about if we have `CLIMT_OPTIMIZE_FLAG` and a user wants the model to run faster including RRTM, so they set that flag to `-O3`, but then it still doesn't speed up RRTM and they either don't notice or they bang their head against a wall for a few hours trying to figure out why RRTM isn't picking up the flag.

If there are parts of RRTM that gain no benefit from optimization, we can find and remove all references to `-OX` in the FFLAGS before inserting them, and then manually specify `-O0`. But I don't know of any such parts.

**Comment by JoyMonteiro** [comment]
If you look at https://github.com/CliMT/climt/blob/develop/climt/_lib/rrtmg_lw/Makefile

you will see that one file (around line 50) is compiled using `-O0` and the rest of the files use `-O3`.

I'm simply suggesting that `FFLAGS` contain no optimization flags, and we define `CLIMT_OPTIMIZE_FLAGS` and use it in places like line 56 and hard code `-O0` in places like line 51.

**Comment by mcgibbon** [comment]
I'm suggesting that FFLAGS contain `-O3` (as appears to be standard in building most applications, and that line 51 of the RRTMG Makefile (and other places) use [find and replace](https://www.gnu.org/software/make/manual/html_node/Text-Functions.html) to force `-O0`.

**Comment by JoyMonteiro** [comment]
but that's not ideal. for instance, if FFLAGS contains some paths required to build some code, and it also contains `-O3`, then for those files which require `-O0` you cannot use FFLAGS even if they require the other options that FFLAGS contains. I have always found it good practice to separate optional flags from required flags.

In setup.py for instance, `-fPIC` is essential, whereas `-O3` is desirable but optional.

**Comment by mcgibbon** [comment]
That's not true. I mean for those that require it we can use find and replace that turns FFLAGS of `-opt -O3 -opt2` to `-opt -O0 -opt2` just for that one file.

**Comment by JoyMonteiro** [comment]
But then how will someone turn optimisation off? since FFLAGS is set in setup.py, there is no way to change the level of optimisation externally. On the other hand, if optimisation is handled by a separate flag, in the script we could check and not modify it if it already exists.

**Comment by mcgibbon** [comment]
You can just as easily ask how someone is supposed to pass any custom FFLAGS currently, or after such a change.

**Comment by JoyMonteiro** [comment]
While it is true that this is not possible currently, I think the correct way to do this would be to allow developers to define a CLIMT_OPTIONAL_FFLAGS. If it is defined, it is appended to FFLAGS. Similarly, a CLIMT_OPTIMIZE_FLAGS can be passed onto downstream Makefiles.

This is standard practice, for example, when you are building a library and need to specify a dependency installed in a non-standard path.

**Comment by mcgibbon** [comment]
I think it makes more sense to have CLIMT_INTERNAL_FFLAGS and have FFLAGS be the user set one, with some reasonable default in setup.py for things like `-O3`.

**Comment by mcgibbon** [comment]
We want to follow standard practice on compiler flags to avoid issues like [this one](https://github.com/numpy/numpy/issues/2122).

**Comment by JoyMonteiro** [comment]
Yes, the developer documentation must specify what flags to set. I'm not sure how numpy is setup, but I don't think it is a good idea to allow the user to set FFLAGS for the climt build. Since FFLAGS is used commonly, I can easily see a scenario where FFLAGS set during a previous compilation get accidentally passed onto a climt build with unexpected results.

The FFLAGS we set in setup.py are expected to always work. If a developer wants to do something unusual, he/she should expect to set an unusual flag, like CLIMT_*

I agree we should have a default optimisation flag like `-O3` which can optionally be overridden by a developer (changing it to `-g` for instance).

**Comment by JoyMonteiro** [comment]
Nearly forgot about this. First of, I will move the optimisation flags out of Makefiles if they are hard-coded. Will get to it next week.
